File: Find the Duplicate Number.java
// https://leetcode.com/problems/find-the-duplicate-number/

/*
- Algorithm is based on Floyd's Tortoise and Hare Cycle Detection
- Time complexity is O(n) while space complexity is O(1)
*/

class Solution {
    public int findDuplicate(int[] nums) {
        
        // If nums only 0 or 1 element, exit loop
        if (nums.length <= 1) {return nums[0];}
        
        // Create 2 variables as tortoise and hare
        // NOTE: Think of this as the starting line
        int tortoise = nums[0];
        int hare = nums[0];
        
        // Traverse array until both animals intersect
        while (true) {
            
            // Move animals
            tortoise = nums[tortoise];
            hare = nums[nums[hare]];
            
            // If both animals intersect, exit loop
            if (tortoise == hare) {break;}
        }
        
        // Move tortoise back to the front of array
        tortoise = nums[0];
        
        // Traverse array again until beginning of cycle is reached
        while (tortoise != hare) {
            
            // Move both animals 1 element at a time
            tortoise = nums[tortoise];
            hare = nums[hare]; // <= Hare is stuck in a cycle, so it will be moving in circles
        }
        
        // Return either animals since both will meet at the beginning of the cycle
        return tortoise;
    }
}

File: Find the Duplicate Number.py
# https://leetcode.com/problems/find-the-duplicate-number/

"""
- Algorithm is based on Floyd's Tortoise and Hare Cycle Detection
- Time complexity is O(n) while space complexity is O(1)
"""

class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        
        # If nums is empty, return 0
        if len(nums) <= 1:
            return 0
        
        # Create 2 pointers as tortoise and hare
        # NOTE: Think of this as the starting line for both animals
        tortoise = nums[0]
        hare = nums[0]
        
        # Traverse list until tortoise and hare intersect
        while True:
            
            # Update positions
            # NOTE: A list should technically be in order. Tortoise is moving one element at a time. Hare is moving based on the appropriate index of the element.
            # 
            # E.g., [2, 1, 4, 3]. At nums[0] == 2, the number 2 should be at the 2nd index or nums[1].
            #       Hare is taking element value 2 and moving to what is suppose to be the appropriate index or nums[1]
            tortoise = nums[tortoise]
            hare = nums[nums[hare]]
            
            # If both animals intersect, exit loop
            if tortoise == hare:
                break
        
        # Send tortoise back to the front of the list
        tortoise = nums[0]
        
        # Traverse list again to find where cycle beings
        while tortoise != hare:
            
            # Move both animals 1 element at a time
            tortoise = nums[tortoise]
            hare = nums[hare] # <= Since Hare is stuck in a cycle, it will go around in a circle
            
        # Return either animals because both are at beginning of the cycle
        return hare
